### ISSUE TITLE
Cannot update credit card info (stripe) or cancel subscription

### ISSUE THREAD

**Issue by OrigamiSamurai** [HabitRPG/habitica#6510]
I am currently unable to update my credit card info for my subscription, or to cancel it.

I have attempted to update the card on a Mac on OS 10.10.5 using Chrome Version 47.0.2526.106 (64-bit) and on my iPhone 6S+ using Safari on iOS 9.2, but I'd guess it has little to do with the browser or device platform.

I tried updating my credit card info for my subscription (I'm using the card directly rather than using PayPal, so I guess that equates to using Stripe). When that didn't work, I attempted to cancel my subscription so I could simply sign up again with the new card. That also failed. I'll start with the info for editing the subscription, then below that I'll include the info for when I tried to cancel.

User ID: 8be6a2a7-1fd5-4caa-b4a9-400ad363d424

---
## EDIT SUBSCRIPTION

I'm getting a 500 Internal Service Error on the post to https://habitica.com/stripe/subscribe/edit
## REQUEST INFO

Headers:

POST /stripe/subscribe/edit HTTP/1.1
Host: habitica.com
Connection: keep-alive
Content-Length: 672
Origin: https://habitica.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36
Content-Type: application/json;charset=UTF-8
Accept: application/json, text/plain, _/_
x-api-user: [REDACTED]
x-client: habitica-web
x-api-key: [REDACTED]
Referer: https://habitica.com/
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.8
Cookie: [REDACTED]

Payload:

{"id":"[string beginning with tok_, REDACTED]","object":"token","card":{"id":"[string beginning with card_, REDACTED]","object":"card","address_city":null,"address_country":null,"address_line1":null,"address_line1_check":null,"address_line2":null,"address_state":null,"address_zip":null,"address_zip_check":null,"brand":"Visa","country":"US","cvc_check":"pass","dynamic_last4":null,"exp_month":[1 digit int, REDACTED],"exp_year":[4 digit year, REDACTED],"funding":"credit","last4":"[string of length 4, REDACTED]","metadata":{},"name":"[my email as a string, REDACTED]","tokenization_method":null,"fingerprint":"[string, REDACTED]","type":"Visa","customer":null},"client_ip":"[IP address as string, REDACTED]","created":[integer, REDACTED],"email":"[my email as a string, REDACTED]","livemode":true,"type":"card","used":false}
## Headers:

HTTP/1.1 500 Internal Server Error
Server: Cowboy
X-Powered-By: Express
X-Ratelimit-Limit: 80
X-Ratelimit-Remaining: 79
Access-Control-Allow-Origin: https://habitica.com
Access-Control-Allow-Methods: OPTIONS,GET,POST,PUT,HEAD,DELETE
Access-Control-Allow-Headers: Content-Type,Accept,Content-Encoding,X-Requested-With,x-api-user,x-api-key
Vary: X-HTTP-Method-Override, Accept-Encoding
Content-Type: application/json; charset=utf-8
Content-Length: 48
Date: Wed, 13 Jan 2016 18:10:45 GMT
Via: 1.1 vegur
## Body:

{"err":"Cannot read property 'id' of undefined"}

---
## CANCEL SUBSCRIPTION

I'm getting a 500 Internal Service Error on the call to https://habitica.com/stripe/subscribe/cancel?_id=8be6a2a7-1fd5-4caa-b4a9-400ad363d424&apiToken=REDACTED
## REQUEST INFO

Headers:

GET /stripe/subscribe/cancel?_id=8be6a2a7-1fd5-4caa-b4a9-400ad363d424&apiToken=REDACTED HTTP/1.1
Host: habitica.com
Connection: keep-alive
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,/;q=0.8
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36
Referer: https://habitica.com/
Accept-Encoding: gzip, deflate, sdch
Accept-Language: en-US,en;q=0.8
Cookie: REDACTED

The request payload is simply the query string parameters of the request, so the _id and my API Token.
## RESPONSE INFO

Headers:

HTTP/1.1 500 Internal Server Error
Server: Cowboy
X-Powered-By: Express X-Ratelimit-Limit: 80
X-Ratelimit-Remaining: 79
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: OPTIONS,GET,POST,PUT,HEAD,DELETE
Access-Control-Allow-Headers: Content-Type,Accept,Content-Encoding,X-Requested-With,x-api-user,x-api-key
Content-Type: application/json; charset=utf-8
Content-Length: 59
Etag: W/"3b-3695263833"
Vary: Accept-Encoding
Date: Tue, 12 Jan 2016 22:54:12 GMT
Via: 1.1 vegur
## Body:

{"err":"Cannot read property 'current_period_end' of null"}

**Comment by Alys** [comment]
The "current_period_end of null" message means that your subscription was cancelled in the past by Stripe, our payment processor company, due to multiple failed payments, however the message was not received by our systems so your subscription in Habitica appeared to continue. Your most recent payment was in April. I've removed the subscription information from your account so you should  now be able to create a new one. If you have any problems with it, please comment again here! I'm going to close this now but we'll still see your messages and we can reopen it if needed.

**Comment by OrigamiSamurai** [comment]
Ok, thanks. I was able to sign up again.

**Comment by charoncherry** [comment]
I'm having this problem as well. I wanted to delete my account so that I could start over from scratch, but I can't delete the account because it has an active subscription. I can't cancel my subscription (which is using a cc that has been cancelled) because of the "current_period_end of null" error.

(I know this issue has been closed, but I thought I'd start here since it is the exact same issue).

User ID: 3273c781-a8ee-4420-a26b-3828d443024c

**Comment by Alys** [comment]
@charoncherry I've confirmed that your subscription payments were cancelled in the Stripe payment processor website and I've removed the subscription information from your Habitica account. 

You will be able to delete your account now if you wish to. However I recommend that you consider other options because your current account has achievements and rewards, such as special pets, that are difficult or impossible to obtain in new accounts. The wiki has a page about various options for restarting: http://habitica.wikia.com/wiki/Start_Over_Options . The Reset option might do what you want. It deletes all your tasks and sets your stats back to those of a new player, but it doesn't remove items that can no longer be obtained easily and doesn't remove anything you might have paid real-world money for.
